File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/6-Graph Representation/Python/directed_graph.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Adjacency list representation of directed graph.

Note that parallel edges are allowed, but not self-loops.
"""

__author__ = 'Ziang Lu'

from graph_basics import AbstractGraph, AbstractVertex


class IllegalArgumentError(ValueError):
    pass


class Vertex(AbstractVertex):
    __slots__ = [
        '_freq_of_emissive_neighbors',
        '_emissive_edges',
        '_freq_of_incident_neighbors',
        '_incident_edges'
    ]

    def __init__(self, vtx_id: int):
        """
        Constructor with parameter.
        :param vtx_id: int
        """
        super().__init__(vtx_id)
        self._freq_of_emissive_neighbors = {}
        self._emissive_edges = []
        self._freq_of_incident_neighbors = {}
        self._incident_edges = []

    def get_emissive_edge_with_head(self, head: AbstractVertex):
        """
        Returns the first emissive edge with the given head.
        :param head: AbstractVertex
        :return: DirectedEdge
        """
        # Check whether the input head is None
        if not head:
            raise IllegalArgumentError('The input head should not be None.')

        for emissive_edge in self._emissive_edges:
            if emissive_edge.head is head:
                return emissive_edge
        # Not found
        return None

    @property
    def emissive_edges(self) -> list:
        """
        Accessor of emissive_edges.
        :return: list[DirectedEdge]
        """
        return self._emissive_edges

    def get_incident_edge_with_tail(self, tail: AbstractVertex):
        """
        Returns the first incident edge with the given tail.
        :param tail: AbstractVertex
        :return: DirectedEdge
        """
        # Check whether the input tail is None
        if not tail:
            raise IllegalArgumentError('The input tail should not be None.')

        for incident_edge in self._incident_edges:
            if incident_edge.tail is tail:
                return incident_edge
        # Not found
        return None

    @property
    def incident_edges(self) -> list:
        """
        Accessor of incident_edges.
        :return: list[DirectedEdge]
        """
        return self._incident_edges

    def add_emissive_edge(self, new_emissive_edge) -> None:
        """
        Adds the given emissive edge to this vertex.
        :param new_emissive_edge: DirectedEdge
        :return: None
        """
        # Check whether the input emissive edge is None
        if not new_emissive_edge:
            raise IllegalArgumentError('The emissive edge to add should not be '
                                       'None.')
        # Check whether the input emissive edge involves this vertex as the tail
        if new_emissive_edge.tail is not self:
            raise IllegalArgumentError('The emissive edge to add should involve'
                                       ' this vertex as the tail.')

        self._emissive_edges.append(new_emissive_edge)

        emissive_neighbor = new_emissive_edge.head
        freq = self._freq_of_emissive_neighbors.get(emissive_neighbor.vtx_id, 0)
        freq += 1
        self._freq_of_emissive_neighbors[emissive_neighbor.vtx_id] = freq

    def add_incident_edge(self, new_incident_edge) -> None:
        """
        Adds the given incident edge to this vertex
        :param new_incident_edge: DirectedEdge
        :return: None
        """
        # Check whether the input incident edge is None
        if not new_incident_edge:
            raise IllegalArgumentError('The incident edge to add should not be '
                                       'None.')
        # Check whether the input incident edge involves this vertex as the head
        if new_incident_edge.head is not self:
            raise IllegalArgumentError('The incident edge to add should involve'
                                       ' this vertex as the head.')

        self._incident_edges.append(new_incident_edge)

        incident_neighbor = new_incident_edge.tail
        freq = self._freq_of_incident_neighbors.get(incident_neighbor.vtx_id, 0)
        freq += 1
        self._freq_of_incident_neighbors[incident_neighbor.vtx_id] = freq

    def remove_emissive_edge(self, emissive_edge_to_remove) -> None:
        """
        Removes the given emissive edge from this vertex.
        :param emissive_edge_to_remove: DirectedEdge
        :return: None
        """
        # Check whether the input emissive edge is None
        if not emissive_edge_to_remove:
            raise IllegalArgumentError(
                'The emissive edge to remove should not be None.'
            )
        # Check whether the input emissive edge involves this vertex as the tail
        if emissive_edge_to_remove.tail is not self:
            raise IllegalArgumentError(
                'The emissive edge to remove should involve this vertex as the '
                'tail.'
            )

        self._emissive_edges.remove(emissive_edge_to_remove)

        emissive_neighbor = emissive_edge_to_remove.head
        freq = self._freq_of_emissive_neighbors.get(emissive_neighbor.vtx_id)
        if freq == 1:
            self._freq_of_emissive_neighbors.pop(emissive_neighbor.vtx_id)
        else:
            freq -= 1
            self._freq_of_emissive_neighbors[emissive_neighbor.vtx_id] = freq

    def remove_incident_edge(self, incident_edge_to_remove) -> None:
        """
        Removes the given incident edge from this vertex.
        :param incident_edge_to_remove: DirectedEdge
        :return: None
        """
        # Check whether the input incident edge is None
        if not incident_edge_to_remove:
            raise IllegalArgumentError(
                'The incident edge to remove should not be None.'
            )
        # Check whether the input incident edge involves this vertex as the head
        if incident_edge_to_remove.head is not self:
            raise IllegalArgumentError(
                'The incident edge to remove should involve this vertex as the '
                'head.'
            )

        self._incident_edges.remove(incident_edge_to_remove)

        incident_neighbor = incident_edge_to_remove.tail
        freq = self._freq_of_incident_neighbors.get(incident_neighbor.vtx_id)
        if freq == 1:
            self._freq_of_incident_neighbors.pop(incident_neighbor.vtx_id)
        else:
            freq -= 1
            self._freq_of_incident_neighbors[incident_neighbor.vtx_id] = freq

    def __repr__(self):
        s = f'Vertex #{self._vtx_id}\n'
        s += f'Its emissive neighbors and frequencies: {self._freq_of_emissive_neighbors}\n'
        s += f'Its incident neighbors and frequencies: {self._freq_of_incident_neighbors}\n'
        return s

    def __eq__(self, other):
        return isinstance(other, Vertex) and self._vtx_id == other.vtx_id


class DirectedEdge(object):
    __slots__ = ['_tail', '_head']

    def __init__(self, tail: Vertex, head: Vertex):
        """
        Constructor with parameter.
        :param tail: Vertex
        :param head: Vertex
        """
        self._tail = tail
        self._head = head

    @property
    def tail(self) -> Vertex:
        """
        Accessor of tail.
        :return: Vertex
        """
        return self._tail

    @property
    def head(self) -> Vertex:
        """
        Accessor of head.
        :return: Vertex
        """
        return self._head

    @tail.setter
    def tail(self, tail: Vertex) -> None:
        """
        Mutator of tail.
        :param tail: Vertex
        :return: None
        """
        self._tail = tail

    @head.setter
    def head(self, head: Vertex) -> None:
        """
        Mutator of head.
        :param head: Vertex
        :return: None
        """
        self._head = head

    def __repr__(self):
        return f'Edge from Vertex #{self._tail.vtx_id} to Vertex #{self._head.vtx_id}'


class DirectedGraph(AbstractGraph):
    __slots__ = []

    def __init__(self):
        """
        Default constructor.
        """
        super().__init__()

    def add_vtx(self, new_vtx_id):
        # Check whether the input vertex is repeated
        if self._find_vtx(new_vtx_id):
            raise IllegalArgumentError('The input vertex is repeated.')

        new_vtx = Vertex(new_vtx_id)
        self._vtx_list.append(new_vtx)

    def _remove_vtx(self, vtx_to_remove):
        # Remove all the edges associated with the vertex to remove
        edges_to_remove = []
        edges_to_remove.extend(vtx_to_remove.emissive_edges)
        edges_to_remove.extend(vtx_to_remove.incident_edges)
        while len(edges_to_remove):
            self._remove_edge(edge_to_remove=edges_to_remove[0])
        # Remove the vertex
        self._vtx_list.remove(vtx_to_remove)

    def add_edge(self, tail_id, head_id):
        # Check whether the input endpoints both exist
        tail, head = self._find_vtx(tail_id), self._find_vtx(head_id)
        if not tail or not head:
            raise IllegalArgumentError("The endpoints don't both exist.")
        # Check whether the input vertices are the same
        if tail_id == head_id:
            raise IllegalArgumentError(
                'The endpoints are the same (self-loop).'
            )

        new_edge = DirectedEdge(tail, head)
        self._add_edge(new_edge=new_edge)

    def _add_edge(self, new_edge):
        tail, head = new_edge.tail, new_edge.head
        tail.add_emissive_edge(new_edge)
        head.add_incident_edge(new_edge)
        self._edge_list.append(new_edge)

    def remove_edge(self, tail_id, head_id):
        # Check whether the input endpoints both exist
        tail, head = self._find_vtx(tail_id), self._find_vtx(head_id)
        if not tail or not head:
            raise IllegalArgumentError("The endpoints don't both exist.")
        # Check whether the edge to remove exists
        edge_to_remove = tail.get_emissive_edge_with_head(head)
        if not edge_to_remove:
            raise IllegalArgumentError("The edge to remove doesn't exist.")

        self._remove_edge(edge_to_remove=edge_to_remove)

    def _remove_edge(self, edge_to_remove):
        tail, head = edge_to_remove.tail, edge_to_remove.head
        tail.remove_emissive_edge(edge_to_remove)
        head.remove_incident_edge(edge_to_remove)
        self._edge_list.remove(edge_to_remove)

File: 4-Shortest Paths Revisit, NP-Complete Problems and What To Do About Them/1-Shortest Paths Revisit/Java/directed_graph/DirectedGraph.java
package directed_graph;

import java.util.ArrayList;
import java.util.LinkedList;

import graph.GraphInterface;

/**
 * Adjacency list representation of a directed graph.
 *
 * Note that parallel edges and self-loops are not allowed.
 * @author Ziang Lu
 */
public class DirectedGraph implements GraphInterface {

    /**
     * Infinity value.
     */
    private static final int INFINITY = 1000000;

    /**
     * Vertex list.
     */
    private final ArrayList<Vertex> vtxList;
    /**
     * Edge list.
     */
    private final ArrayList<DirectedEdge> edgeList;
    /**
     * Bellman-Ford subproblem solutions.
     * Since there are only O(n^2) subproblems, the first time we solve a
     * subproblem, we can cache its solution in a global take for O(1) lookup
     * time later on.
     */
    private Integer[][] bellmanFordSubproblems;

    /**
     * Default constructor.
     */
    public DirectedGraph() {
        vtxList = new ArrayList<>();
        edgeList = new ArrayList<>();
    }

    @Override
    public void addVtx(int newVtxID) {
        // Check whether the input vertex is repeated
        if (findVtx(newVtxID) != null) {
            throw new IllegalArgumentException("The input vertex is repeated.");
        }

        Vertex newVtx = new Vertex(newVtxID);
        vtxList.add(newVtx);
    }

    /**
     * Private helper method to find the given vertex in this adjacency list.
     * @param vtxID vertex ID to look for
     * @return vertex if found, null if not found
     */
    private Vertex findVtx(int vtxID) {
        for (Vertex vtx : vtxList) {
            if (vtx.id() == vtxID) {
                return vtx;
            }
        }
        // Not found
        return null;
    }

    @Override
    public void removeVtx(int vtxID) {
        // Check whether the input vertex exists
        Vertex vtxToRemove = findVtx(vtxID);
        if (vtxToRemove == null) {
            throw new IllegalArgumentException("The inpue vertex doesn't exist.");
        }

        removeVtx(vtxToRemove);
    }

    /**
     * Private helper method to remove the given vertex from this graph.
     * @param vtxToRemove vertex to remove
     */
    private void removeVtx(Vertex vtxToRemove) {
        // Remove all the edges associated with the vertex to remove
        ArrayList<DirectedEdge> edgesToRemove = new ArrayList<>();
        edgesToRemove.addAll(vtxToRemove.emissiveEdges());
        edgesToRemove.addAll(vtxToRemove.incidentEdges());
        while (edgesToRemove.size() > 0) {
            DirectedEdge edgeToRemove = edgesToRemove.get(0);
            removeEdge(edgeToRemove);
        }
        // Remove the vertex
        vtxList.remove(vtxToRemove);
    }

    @Override
    public void addEdge(int tailID, int headID, int length) {
        // Check whether the input endpoints both exist
        Vertex tail = findVtx(tailID), head = findVtx(headID);
        if ((tail == null) || (head == null)) {
            throw new IllegalArgumentException("The endpoints don't both exist.");
        }
        // Check whether the input endpoints both exist
        if (tailID == headID) {
            throw new IllegalArgumentException("The endpoints are the same (self-loop).");
        }

        DirectedEdge newEdge = new DirectedEdge(tail, head, length);
        addEdge(newEdge);
    }

    /**
     * Private helper method to add the given edge to this graph.
     * @param newEdge new edge
     */
    private void addEdge(DirectedEdge newEdge) {
        Vertex tail = newEdge.tail(), head = newEdge.head();
        tail.addEmissiveEdge(newEdge);
        head.addIncidentEdge(newEdge);
        edgeList.add(newEdge);
    }

    @Override
    public void removeEdge(int tailID, int headID) {
        // Check whether the input endpoints both exist
        Vertex tail = findVtx(tailID), head = findVtx(headID);
        if ((tail == null) || (head == null)) {
            throw new IllegalArgumentException("The endpoints don't both exist.");
        }
        // Check whether the edge to remove exists
        DirectedEdge edgeToRemove = tail.getEmissiveEdgeWithHead(head);
        if (edgeToRemove == null) {
            throw new IllegalArgumentException("The edge to remove doesn't exist.");
        }

        removeEdge(edgeToRemove);
    }

    /**
     * Private helper method to remove the given edge from this graph.
     * @param edgeToRemove edge to remove
     */
    private void removeEdge(DirectedEdge edgeToRemove) {
        Vertex tail = edgeToRemove.tail(), head = edgeToRemove.head();
        tail.removeEmissiveEdge(edgeToRemove);
        head.removeIncidentEdge(edgeToRemove);
        edgeList.remove(edgeToRemove);
    }

    /**
     * Removes all the directed edges between a vertex pair from this graph.
     * @param tailID tail ID
     * @param headID head ID
     */
    public void removeDirectedEdgesBetweenPair(int tailID, int headID) {
        try {
            removeEdge(tailID, headID);
        } catch (IllegalArgumentException ex) {};
    }

    @Override
    public void showGraph() {
        System.out.println("The vertices are:");
        for (Vertex vtx : vtxList) {
            System.out.println(vtx);
        }
        System.out.println("The edges are:");
        for (DirectedEdge edge : edgeList) {
            System.out.println(edge);
        }
    }

    @Override
    public ArrayList<LinkedList<Integer>> bellmanFordShortestPaths(int srcVtxID) {
        // Check whether the input source vertex exists
        Vertex srcVtx = findVtx(srcVtxID);
        if (srcVtx == null) {
            throw new IllegalArgumentException("The source vertex doesn't exist.");
        }

        int n = vtxList.size();
        // Initialization
        bellmanFordSubproblems = new Integer[n][n];
        for (Vertex vtx : vtxList) {
            if (vtx == srcVtx) {
                bellmanFordSubproblems[vtx.id()][0] = 0;
            } else {
                bellmanFordSubproblems[vtx.id()][0] = INFINITY;
            }
        }
        // Bottom-up calculation
        for (int budget = 1; budget <= (n - 1); ++budget) {
            for (Vertex vtx : vtxList) {
                // Case 1: P(s, v, i) has <= (i - 1) edges (i.e., P doesn't use up all of its budget i.)
                Integer minPathLength = bellmanFordSubproblems[vtx.id()][budget - 1];
                // Case 2: P(s, v, i) has exactly i edges (i.e., P uses up all of its budget i.), with final hop (w, v)
                for (DirectedEdge incidentEdge : vtx.incidentEdges()) {
                    Vertex w = incidentEdge.tail();
                    // By plucking off the final hop (w, v), we form P'(s, w, i - 1).
                    int pathLength = bellmanFordSubproblems[w.id()][budget - 1] + incidentEdge.length();
                    if (pathLength < minPathLength) {
                        minPathLength = pathLength;
                    }
                }
                // P(s, v, i) is the minimum among the above (1 + in-degree(v)) candidates.
                bellmanFordSubproblems[vtx.id()][budget] = minPathLength;
            }
        }
        /*
         * Extension to detect negative cycles reachable from s:
         * If the input graph has negative cycles reachable from s, then we just run the outer loop for one extra
         * iteration, and check if there is still an improvement on some vertex. If so, then the input graph must have
         * negative cycles reachable from s.
         */
        boolean madeUpdateInExtraIter = false;
        for (Vertex vtx : vtxList) {
            Integer minPathLength = bellmanFordSubproblems[vtx.id()][n - 1];
            for (DirectedEdge incidentEdge : vtx.incidentEdges()) {
                Vertex w = incidentEdge.tail();
                int pathLength = bellmanFordSubproblems[w.id()][n - 1] + incidentEdge.length();
                if (pathLength < minPathLength) {
                    minPathLength = pathLength;
                    madeUpdateInExtraIter = true;
                }
            }
        }
        if (madeUpdateInExtraIter) {
            throw new IllegalArgumentException("The graph has negative cycles reachable from the source vertex.");
        }
        // The final solution lies in exactly subproblems[v][n - 1].
        return reconstructShortestPaths();
        // Outer for-loop: n iterations
        // Inner for-loop: sum(in-degree(v)) = m
        // => Overall running time complexity: O(mn)
        // Overall space complexity: O(n^2)
    }

    /**
     * Private helper method to reconstruct the shortest paths according to the
     * optimal solution using backtracking.
     * @return shortest paths
     */
    private ArrayList<LinkedList<Integer>> reconstructShortestPaths() {
        ArrayList<LinkedList<Integer>> shortestPaths = new ArrayList<>();
        for (Vertex vtx : vtxList) {
            LinkedList<Integer> shortestPath = new LinkedList<>();
            shortestPath.add(vtx.id());
            Vertex currVtx = vtx;
            int budget = vtxList.size() - 1;
            while (budget >= 1) {
                // Find the previous vertex to backtrack
                Vertex prevVtx = currVtx;
                Integer minPathLength = bellmanFordSubproblems[currVtx.id()][budget - 1];
                for (DirectedEdge incidentEdge : currVtx.incidentEdges()) {
                    Vertex w = incidentEdge.tail();
                    int pathLength = bellmanFordSubproblems[w.id()][budget - 1] + incidentEdge.length();
                    if (pathLength < minPathLength) {
                        prevVtx = w;
                        minPathLength = pathLength;
                    }
                }
                if (prevVtx != currVtx) {
                    shortestPath.addFirst(prevVtx.id());
                }
                currVtx = prevVtx;
                --budget;
            }
            shortestPaths.add(shortestPath);
        }
        return shortestPaths;
        // Running time complexity: O(mn)
    }

    @Override
    public ArrayList<LinkedList<Integer>> bellmanFordShortestPathsOptimized(int srcVtxID) {
        // Check whether the input source vertex exists
        Vertex srcVtx = findVtx(srcVtxID);
        if (srcVtx == null) {
            throw new IllegalArgumentException("The source vertex doesn't exist.");
        }

        int n = vtxList.size();
        // Initialization
        // Space optimization: We only keep track of the subproblem solutions in the previous outer iteration.
        int[] prevIterSubproblems = new int[n], currIterSubproblems = new int[n];
        for (Vertex vtx : vtxList) {
            if (vtx != srcVtx) {
                prevIterSubproblems[vtx.id()] = INFINITY;
            }
        }
        // In order to recover the ability to reconstruct the shortest paths, we also keep track of the penultimate
        // vertices in the previous outer iteration.
        Vertex[] prevIterPenultimateVtxs = new Vertex[n], currIterPenultimateVtxs = new Vertex[n];
        // Bottom-up calculation
        int budget = 1;
        // Optimization: Early-stopping
        // The algorithm may stop early when in the current iteration, no update is made for any vertex.
        boolean madeUpdateInIter = true;
        while ((budget <= (n - 1)) && madeUpdateInIter) {
            madeUpdateInIter = false;
            for (Vertex vtx : vtxList) {
                Integer minPathLength = prevIterSubproblems[vtx.id()];
                Vertex penultimateVtx = prevIterPenultimateVtxs[vtx.id()];
                for (DirectedEdge incidentEdge : vtx.incidentEdges()) {
                    Vertex w = incidentEdge.tail();
                    int pathLength = prevIterSubproblems[w.id()] + incidentEdge.length();
                    if (pathLength < minPathLength) {
                        minPathLength = pathLength;
                        madeUpdateInIter = true;
                        penultimateVtx = w;
                    }
                }
                currIterSubproblems[vtx.id()] = minPathLength;
                currIterPenultimateVtxs[vtx.id()] = penultimateVtx;
            }
            ++budget;
            System.arraycopy(currIterSubproblems, 0, prevIterSubproblems, 0, n);
            System.arraycopy(currIterPenultimateVtxs, 0, prevIterPenultimateVtxs, 0, n);
        }
        madeUpdateInIter = false;
        for (Vertex vtx : vtxList) {
            Integer minPathLength = prevIterSubproblems[vtx.id()];
            for (DirectedEdge incidentEdge : vtx.incidentEdges()) {
                Vertex w = incidentEdge.tail();
                int pathLength = prevIterSubproblems[w.id()] + incidentEdge.length();
                if (pathLength < minPathLength) {
                    minPathLength = pathLength;
                    madeUpdateInIter = true;
                }
            }
            currIterSubproblems[vtx.id()] = minPathLength;
        }
        if (madeUpdateInIter) {
            throw new IllegalArgumentException("The graph has negative cycles reachable from the source vertex.");
        }
        // The final solution lies in exactly prevIterSubproblems.

        // We can reconstruct the shortest paths from these penultimate vertices.
        return reconstructShortestPathsOptimized(prevIterPenultimateVtxs);
        // Overall running time complexity: O(mn)
        // Overall space complexity: O(n)
    }

    /**
     * Private helper method to reconstruct the shortest paths according to the
     * penultimate vertices in the shortest paths using backtracking.
     * @param penultimateVtxs penultimate vertices in the shortest paths
     * @return shortest paths
     */
    private ArrayList<LinkedList<Integer>> reconstructShortestPathsOptimized(Vertex[] penultimateVtxs) {
        ArrayList<LinkedList<Integer>> shortestPaths = new ArrayList<>();
        for (Vertex vtx : vtxList) {
            LinkedList<Integer> shortestPath = new LinkedList<>();
            Vertex currVtx = vtx;
            while (currVtx != null) {
                shortestPath.addFirst(currVtx.id());
                Vertex prevVtx = penultimateVtxs[currVtx.id()];
                currVtx = prevVtx;
            }
            shortestPaths.add(shortestPath);
        }
        return shortestPaths;
        // Running time complexity: O(n^2)
    }

    @Override
    public ArrayList<LinkedList<Integer>> bellmanFordShortestPathsDestDriven(int destVtxID) {
        // Check whether the input destination vertex exists
        Vertex destVtx = findVtx(destVtxID);
        if (destVtx == null) {
            throw new IllegalArgumentException("The destination vertex doesn't exist.");
        }

        int n = vtxList.size();
        // Initialization
        bellmanFordSubproblems = new Integer[n][n];
        for (Vertex vtx : vtxList) {
            if (vtx == destVtx) {
                bellmanFordSubproblems[vtx.id()][0] = 0;
            } else {
                bellmanFordSubproblems[vtx.id()][0] = INFINITY;
            }
        }
        // Bottom-up calculation
        for (int budget = 1; budget <= (n - 1); ++budget) {
            for (Vertex vtx : vtxList) {
                // Case 1: P(v, d, i) has <= (i - 1) edges (i.e., P doesn't use up all of its budget i.)
                Integer minPathLength = bellmanFordSubproblems[vtx.id()][budget - 1];
                // Case 2: P(v, d, i) has exactly i edges (i.e., P uses up all of its budget i.), with first hop (v, w)
                for (DirectedEdge emissiveEdge : vtx.emissiveEdges()) {
                    Vertex w = emissiveEdge.head();
                    // By plucking off the first hop (v, w), we form P'(w, d, i - 1).
                    int pathLength = bellmanFordSubproblems[w.id()][budget - 1] + emissiveEdge.length();
                    if (pathLength < minPathLength) {
                        minPathLength = pathLength;
                    }
                }
                // P(v, d, i) is the minimum among the above (1 + out-degree(v)) candidates.
                bellmanFordSubproblems[vtx.id()][budget] = minPathLength;
            }
        }
        /*
         * Extension to detect negative cycles reachable from d:
         * If the input graph has negative cycles reachable from d, then we just run the outer loop for one extra
         * iteration, and check if there is still an improvement on some vertex. If so, then the input graph must have
         * negative cycles reachable from d.
         */
        boolean madeUpdateInExtraIter = false;
        for (Vertex vtx : vtxList) {
            Integer minPathLength = bellmanFordSubproblems[vtx.id()][n - 1];
            for (DirectedEdge emissiveEdge : vtx.emissiveEdges()) {
                Vertex w = emissiveEdge.head();
                int pathLength = bellmanFordSubproblems[w.id()][n - 1] + emissiveEdge.length();
                if (pathLength < minPathLength) {
                    minPathLength = pathLength;
                    madeUpdateInExtraIter = true;
                }
            }
        }
        if (madeUpdateInExtraIter) {
            throw new IllegalArgumentException("The graph has negative cycles reachable from the destination vertex.");
        }
        // The final solution lies in exactly subproblems[v][n - 1].
        return reconstructShortestPathsDestDriven();
        // Outer for-loop: n iterations
        // Inner for-loop: sum(out-degree(v)) = m
        // Overall running time complexity: O(mn)
        // Overall space complexity: O(n^2)
    }

    /**
     * Private helper method to reconstruct the destination-driven shortest
     * paths according to the optimal solution using backtracking.
     * @return destination-driven shortest paths
     */
    private ArrayList<LinkedList<Integer>> reconstructShortestPathsDestDriven() {
        ArrayList<LinkedList<Integer>> shortestPaths = new ArrayList<>();
        for (Vertex vtx : vtxList) {
            LinkedList<Integer> shortestPath = new LinkedList<>();
            shortestPath.add(vtx.id());
            Vertex currVtx = vtx;
            int budget = vtxList.size() - 1;
            while (budget >= 1) {
                // Find the next vertex
                Vertex nextVtx = currVtx;
                Integer minPathLength = bellmanFordSubproblems[vtx.id()][budget - 1];
                for (DirectedEdge emissiveEdge : currVtx.emissiveEdges()) {
                    Vertex w = emissiveEdge.head();
                    int pathLength = bellmanFordSubproblems[w.id()][budget - 1] + emissiveEdge.length();
                    if (pathLength < minPathLength) {
                        nextVtx = w;
                        minPathLength = pathLength;
                    }
                }
                if (nextVtx != currVtx) {
                    shortestPath.add(nextVtx.id());
                }
                currVtx = nextVtx;
                --budget;
            }
            shortestPaths.add(shortestPath);
        }
        return shortestPaths;
        // Running time complexity: O(mn)
    }

    @Override
    public ArrayList<LinkedList<Integer>> bellmanFordShortestPathsDestDrivenOptimized(int destVtxID) {
        // Check whether the input destination vertex exists
        Vertex destVtx = findVtx(destVtxID);
        if (destVtx == null) {
            throw new IllegalArgumentException("The destination vertex doesn't exist.");
        }

        int n = vtxList.size();
        // Initialization
        // Space optimization: We only keep track of the subproblem solutions in the previous outer iteration.
        int[] prevIterSubproblems = new int[n], currIterSubproblems = new int[n];
        for (Vertex vtx : vtxList) {
            if (vtx != destVtx) {
                prevIterSubproblems[vtx.id()] = INFINITY;
            }
        }
        // In order to recover the ability to reconstruct the shortest paths, we also keep track of the next vertices in
        // the previous outer iteration.
        Vertex[] prevIterNextVtxs = new Vertex[n], currIterNextVtxs = new Vertex[n];
        // Bottom-up calculation
        int budget = 1;
        // Optimization: Early-stopping
        // The algorithm may stop early when in the current iteration, no update is made for any vertex.
        boolean madeUpdateInIter = true;
        while ((budget <= (n - 1)) && madeUpdateInIter) {
            madeUpdateInIter = false;
            for (Vertex vtx : vtxList) {
                Integer minPathLength = prevIterSubproblems[vtx.id()];
                Vertex nextVtx = prevIterNextVtxs[vtx.id()];
                for (DirectedEdge emissiveEdge : vtx.emissiveEdges()) {
                    Vertex w = emissiveEdge.head();
                    int pathLength = prevIterSubproblems[w.id()] + emissiveEdge.length();
                    if (pathLength < minPathLength) {
                        minPathLength = pathLength;
                        madeUpdateInIter = true;
                        nextVtx = w;
                    }
                }
                currIterSubproblems[vtx.id()] = minPathLength;
                currIterNextVtxs[vtx.id()] = nextVtx;
            }
            ++budget;
            System.arraycopy(currIterSubproblems, 0, prevIterSubproblems, 0, n);
            System.arraycopy(currIterNextVtxs, 0, prevIterNextVtxs, 0, n);
        }
        madeUpdateInIter = false;
        for (Vertex vtx : vtxList) {
            Integer minPathLength = prevIterSubproblems[vtx.id()];
            for (DirectedEdge emissiveEdge : vtx.emissiveEdges()) {
                Vertex w = emissiveEdge.head();
                int pathLength = prevIterSubproblems[w.id()] + emissiveEdge.length();
                if (pathLength < minPathLength) {
                    minPathLength = pathLength;
                    madeUpdateInIter = true;
                }
            }
            currIterSubproblems[vtx.id()] = minPathLength;
        }
        if (madeUpdateInIter) {
            throw new IllegalArgumentException("The graph has negative cycles reachable from the destination vertex.");
        }
        // The final solution lies in exactly in prevIterSubprblems.
        return reconstructShortestPathsDestDrivenOptimized(prevIterNextVtxs);
        // Overall running time complexity: O(mn)
        // Overall space complexity: O(n)
    }

    /**
     * Private helper method to reconstruct the destination-driven shortest
     * paths according to the next vertices in the shortest paths using
     * backtracking.
     * @param nextVtxs next vertices in the shortest paths
     * @return destination-driven shortest paths
     */
    private ArrayList<LinkedList<Integer>> reconstructShortestPathsDestDrivenOptimized(Vertex[] nextVtxs) {
        ArrayList<LinkedList<Integer>> shortestPaths = new ArrayList<>();
        for (Vertex vtx : vtxList) {
            LinkedList<Integer> shortestPath = new LinkedList<>();
            Vertex currVtx = vtx;
            while (currVtx != null) {
                shortestPath.add(currVtx.id());
                Vertex nextVtx = nextVtxs[currVtx.id()];
                currVtx = nextVtx;
            }
            shortestPaths.add(shortestPath);
        }
        return shortestPaths;
        // Running time complexity: O(n^2)
    }

    @Override
    public ArrayList<LinkedList<Integer>> shortestPathsDestDrivenPushBased(int destVtxID) {
        // Check whether the input destination vertex exists
        Vertex destVtx = findVtx(destVtxID);
        if (destVtx == null) {
            throw new IllegalArgumentException("The destination vertex doesn't exist.");
        }

        int n = vtxList.size();
        // Initialization
        int[] minPathLengths = new int[n];
        for (Vertex vtx : vtxList) {
            if (vtx != destVtx) {
                minPathLengths[vtx.id()] = INFINITY;
            }
        }
        Vertex[] nextVtxs = new Vertex[n];
        // Start the notifications from the destination vertex
        for (DirectedEdge incidentEdge : destVtx.incidentEdges()) {
            notifyTail(incidentEdge, minPathLengths, nextVtxs);
        }
        return reconstructShortestPathsDestDrivenOptimized(nextVtxs);
        // Overall running time complexity: O(2^n)
    }

    /**
     * Private helper method to notify the tail of the given edge with an
     * updated minimum path length of the head of the given edge recursively.
     * @param edge given edge
     * @param minPathLengths destination-driven shortest distances
     * @param nextVtxs next vertices in the shortest paths
     */
    private void notifyTail(DirectedEdge edge, int[] minPathLengths, Vertex[] nextVtxs) {
        Vertex nextVtx = edge.head(), currVtx = edge.tail();
        int updatedPathLength = minPathLengths[nextVtx.id()];
        int newPathLength = updatedPathLength + edge.length();
        if (newPathLength < minPathLengths[currVtx.id()]) {
            minPathLengths[currVtx.id()] = newPathLength;
            nextVtxs[currVtx.id()] = nextVtx;
            // Notify all incident neighbors
            for (DirectedEdge incidentEdge : currVtx.incidentEdges()) {
                notifyTail(incidentEdge, minPathLengths, nextVtxs);
            }
        }
    }

}
